man/man8/probe.8
.TH PROBE 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/probe.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: probe.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
probe \- Generates a Probe
.SH SYNOPSIS
.in +.5i
.ti -.5i
probe 	
.in -.5i
.in +.4i
\%[-t [Recipients ...]] 
.br
\%[-s Message-size] 
.br
\%[-e Encoded-Information-Types]
.br
\%[-u UA-id] 
.br
\%[-i]
.br
\%[-a]
.in -.4i
.SH DESCRIPTION
.I Probe,
generates a message probe which should always return either a 
successful or a failed Delivery Report.
.PP
It has the following options
.TP
\-t
Interpret the remaining arguments up to the next switch as
\*(lqTo:\*(rq Recipients.
.TP
\-s
Specifies the size of the message that will be sent later on.
.TP
\-e 
Specifies one or more encoded information types within the message 
to be sent later. The list of types should be seperated 
by a \*(lq,\*(rq and are found in the PP Documentation Volume 1. 
.TP
\-u
Specifies a User id. This should be a maximum of 16 characters.
.TP
\-i
This option requests implicit conversion.
.TP
\-a 
This option allows for alternate recipients.

.SH "SEE ALSO"
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Tools/probe/probe.c
/* probe.c: Generates a probe */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Tools/probe/RCS/probe.c,v 6.0 1991/12/18 20:32:41 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Tools/probe/RCS/probe.c,v 6.0 1991/12/18 20:32:41 jpo Rel $
 *
 * $Log: probe.c,v $
 * Revision 6.0  1991/12/18  20:32:41  jpo
 * Release 6.0
 *
 */



#include "head.h"
#include "prm.h"
#include "q.h"
#include <isode/cmd_srch.h>



/* --- externals --- */
extern char				*loc_dom_site,
					*local_822_chan,
					*cont_822;
extern ADDR				*make_adr_new();


/* --- queue variables --- */
static struct prm_vars                  PRMstruct;
static Q_struct                         Qstruct;
static struct prm_vars                  *PrmPtr = &PRMstruct;
Q_struct                                *QuePtr = &Qstruct;


/* --- other statics  --- */
static char				*myname = NULLCP;

#define	OPT_TO				1
#define OPT_SIZE			2
#define	OPT_EIT				3
#define	OPT_UAID			4
#define	OPT_IMPCONV			5
#define OPT_ALTRECIP			6
#define OPT_DEBUG			7

static int				param = OPT_TO;

CMD_TABLE  probe_options[] = { /* probe commandline options */
	"-t",				OPT_TO,
	"-s",				OPT_SIZE,
	"-e",				OPT_EIT,
	"-u",				OPT_UAID,
	"-i",				OPT_IMPCONV,
	"-a",				OPT_ALTRECIP,
	0,				-1
	};



static void		probe_set_rest(), probe_checks(), usage();	





/* ---------------------  Begin  Routines  -------------------------------- */




main (argc, argv)
int     argc;
char    **argv;
{
	char	*filename;
	ADDR	*ap;
	int	retval;


	bzero ((char *)PrmPtr, sizeof(*PrmPtr));
	bzero ((char *)QuePtr, sizeof(*QuePtr));
	(void) set_from();

	myname = argv[0];

	if (argc > 16)
		usage();

	if (argc == 1) {
		(void) prompt_probe_stdin();	
		probe_process();
	}
		

	while (--argc > 0) {
		argv ++;
		if (**argv == '-')
			switch (param = cmd_srch (*argv, probe_options)) {

			case OPT_TO:
			case OPT_SIZE:
			case OPT_EIT:
			case OPT_UAID:
				break;

			case OPT_IMPCONV:
				QuePtr -> implicit_conversion_prohibited = YES;
				break;
			case OPT_ALTRECIP:
				QuePtr->alternate_recip_allowed = YES;
				break;
			default:
                        	printf ("\n***Error: Unknown option '%s'\n",
					*argv);
                        	usage();
			}   /* -- end of switch -- */

		else {
			switch (param) {
			case OPT_TO:
				ap = make_adr_new (*argv, AD_RECIPIENT);
                		adr_add (&QuePtr -> Raddress, ap);
				break;
			case OPT_SIZE:
				QuePtr -> msgsize = atoi(*argv);
				break;
			case OPT_EIT:
				retval = set_encoded (*argv);
				if (retval == NOTOK) {
                        		printf ("\n***Error: eits! '%s'\n",
					*argv);
                        		usage();
				}
				break;
			case OPT_UAID:
				QuePtr -> ua_id = strdup (*argv);
				break;
			default:
                        	printf ("\n***Error: Unknown option '%s'\n",
					*argv);
				usage();

			}  /* -- end of switch -- */

		}   /* -- end of else -- */

	}  /* -- end of while -- */

	
	probe_process();
}






/* ---------------------  Static  Routines  ------------------------------- */






static void usage()
{
	printf ("\n\n");
	printf ("Usage:   %s\n", myname); 
	printf ("	  	[-t Recipient]\n");
	printf ("		[-s MessageSize]\n");
	printf ("	  	[-e EncodedInfoTypes]\n");
	printf ("		[-u UAid]\n");
	printf ("	  	[-i ImplicitConversion]\n");
	printf ("	  	[-a AlternateRecipientAllowed]\n");
	printf ("\n\n");
	exit (1);
}





static int probe_process()
{
	int	retval; 

	(void) probe_checks();

	sys_init (myname);
	if (or_init() == NOTOK)	 exit (1);
	or_myinit();
	(void) probe_set_rest();

	printf ("Processing ...\n");
	retval = probe_submit();

	if (rp_isbad (retval)) {
		printf ("*** Error: Unable to submit ***\n\n");
		exit (1);
	}
		
	exit (0);
}




static void probe_set_rest()
{
	prm_init (PrmPtr);
        PrmPtr->prm_opts = PRM_ACCEPTALL;

	QuePtr -> msgtype	= MT_PMPDU;
	QuePtr -> cont_type	= strdup(cont_822);
	QuePtr -> priority	= PRIO_URGENT;
	QuePtr -> inbound 	= list_rchan_new (loc_dom_site, local_822_chan);

	if (QuePtr -> msgsize == 0)
	QuePtr -> msgsize 	= 500;

	if (QuePtr -> encodedinfo.eit_types == 0)
		if (set_encoded (NULLCP) == NOTOK) {
			printf ("\n***Error: Unable to set eit\n");
			exit (1);
		}
	(void) MPDUid_new (&QuePtr -> msgid);
}




static void probe_checks()
{

	ADDR	*ap;
	int	err = FALSE;

        if (isstr (QuePtr -> ua_id) && (int)strlen (QuePtr -> ua_id) > 16) {
                printf ("*** Error:  Invalid UA size (max 16) '%s' ***\n",
                        QuePtr -> ua_id);
               	exit (1);
	}

        if (QuePtr -> msgsize < 0) {
                printf ("*** Error:  Invalid Message Size '%d' ***\n",
                        QuePtr -> msgsize);
               	exit (1);
	}
}




static int probe_submit()
{
	RP_Buf          rp;
	ADDR            *ap;
	int		submit_status = NOTOK;

	if (rp_isbad (io_init(rp))) {
		printf ("*** Error: probe_submit:  io_init\n");
		exit (1);
	}

	if (rp_isbad (io_wprm (PrmPtr, &rp))) {
		printf ("*** Error: probe_submit:  io_wprm\n");
		goto  probe_submit_end;
	}

	if (rp_isbad (io_wrq (QuePtr, &rp))) {
		printf ("*** Error: probe_submit:  io_wrq\n");
		goto  probe_submit_end;
	}

	if (rp_isbad (io_wadr (QuePtr->Oaddress, AD_ORIGINATOR, &rp))) {
		printf ("*** Error: probe_submit:  io_wadr ('%s')\n",
			QuePtr->Oaddress->ad_value);
		goto  probe_submit_end;
	}


	for (ap = QuePtr->Raddress; ap; ap = ap->ad_next) {

		/* -- Red Book X.411 Clause 4.3.2.2 -- */ 
		ap -> ad_mtarreq = AD_MTA_CONFIRM;
		ap -> ad_usrreq = AD_USR_CONFIRM;

		if (rp_isbad (io_wadr (ap, AD_RECIPIENT, &rp))) {
			printf ("*** Error: probe_submit:  io_wadr ('%s')\n",
				ap->ad_value);
			goto  probe_submit_end;
		}
	}


	if (rp_isbad (io_adend (&rp))) {
		printf ("*** Error: probe_submit:  io_adend\n");
		goto  probe_submit_end;
	}
	else printf ("%s\n", rp.rp_line);

	submit_status = OK;

probe_submit_end: ;
	io_end (submit_status);
	return submit_status;
}
